File: java/src/Coupon.java
// Coupon.java
// Represents a swap cash flow
import java.util.Date;
import java.util.ArrayList;

import org.apache.commons.math3.analysis.UnivariateFunction;
import org.apache.commons.math3.analysis.interpolation.SplineInterpolator;
import org.apache.commons.math3.analysis.interpolation.UnivariateInterpolator;
import org.apache.commons.math3.analysis.polynomials.PolynomialSplineFunction;

public class Coupon extends Period {
	private DaycountConvention _dayCount;
	private double _nbYears;
    private Double _userForwardRate;
    private double _startDiscountFactor;
    private double _endDiscountFactor;
    private double _cashFlow;

    public Coupon(Date startDate, Date endDate)
    {
    	this(startDate, endDate, DaycountConvention.ActualActual); 
    }

    public Coupon(Date startDate, Date endDate, DaycountConvention dayCount)
    {
    	setStartDate(startDate);
    	setEndDate(endDate);
    	setDayCount(dayCount); 
    }

    public long getNbDays()
    {
    	return DateUtils.CountDays(this.getStartDate(), this.getEndDate());
	}

    private void setDayCount(DaycountConvention dayCount)
    {
    	_dayCount = dayCount;
	}

    public DaycountConvention getDayCount()
    {
    	return _dayCount != null ? _dayCount : DaycountConvention.ActualActual;
	}

    public double getNbYears()
    {
    	_nbYears = DateUtils.YearFrac(this.getStartDate(), this.getEndDate(), this.getDayCount());
    	return _nbYears;
	}

    public void setStartDiscountFactor(double df)
    {
    	_startDiscountFactor = df;
    }
    
    public double getStartDiscountFactor()
    {
    	return _startDiscountFactor;
    }

    public void setEndDiscountFactor(double df)
    {
    	_endDiscountFactor = df;
    }
    
    public double getEndDiscountFactor()
    {
    	return _endDiscountFactor;
    }

    public void setForwardRate(double rate)
    {
    	_userForwardRate = rate;	
    }
    
    public double getForwardRate()
    {
    	if (!(_userForwardRate == null)) return _userForwardRate.doubleValue();
        return (getStartDiscountFactor() == 0 || getNbYears() == 0) ? 0 : ((1 / (getEndDiscountFactor() / getStartDiscountFactor())) - 1) / getNbYears();
    }
    
    public double getForwardDiscountFactor()
    {
        return (getStartDiscountFactor() == 0) ? 0 : (getEndDiscountFactor() / getStartDiscountFactor());
    }

    public void setCashFlow(double cf)
    {
    	_cashFlow = cf;
    }
    
    public double getCashFlow()
    {
    	return _cashFlow;
    }

    public void CalculateDiscountFactor(ArrayList<ZeroCouponDataPoint> discountCurve)
    {
    	UnivariateInterpolator interp = new SplineInterpolator();
    	double[] xArray = discountCurve.stream().mapToDouble(x -> x.getEndDate().getTime()).toArray();
		double[] yArray = discountCurve.stream().mapToDouble(y -> y.getDiscountFactor()).toArray();
		UnivariateFunction cs = interp.interpolate(xArray, yArray);
        setStartDiscountFactor(cs.value(getStartDate().getTime()));
        setEndDiscountFactor(cs.value(getEndDate().getTime()));
    }

    public void CalculateCashFlow(double notional)
    {
    	CalculateCashFlow(notional, 0.0, false);
    }

    public void CalculateCashFlow(double notional, double spread)
    {
    	CalculateCashFlow(notional, spread, false);
    }

    public void CalculateCashFlow(double notional, double spread, boolean maturity)
    {
        double couponRate = getForwardRate() + spread;
        double redemption = maturity ? 1.0 : 0.0;
        double nbYears = getNbYears();
        setCashFlow((redemption + (couponRate * nbYears)) * notional);
    }

}

File: csharp/vanilla_swap/vanilla_swap/Coupon.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using mn = MathNet.Numerics.Interpolation;

namespace VanillaSwap
{
    public class Coupon
    {
        private double? _userForwardRate;

        public Coupon(DateTime startDate, DateTime endDate, DayCountBasis dayCount = DayCountBasis.ActualActual)
        {
            StartDate = startDate;
            EndDate = endDate;
            NbDays = (EndDate - StartDate).Days;
            DayCount = dayCount;
            NbYears = DateUtils.YearFrac(StartDate, EndDate);
        }

        public DateTime StartDate { get; set; }

        public DateTime EndDate { get; set; }

        public int NbDays { get; set; }

        public DayCountBasis DayCount { get; set; }

        public double NbYears { get; set; }

        public double StartDiscountFactor { get; set; }

        public double EndDiscountFactor { get; set; }

        public double ForwardRate {
            get
            {
                if (_userForwardRate.HasValue) return _userForwardRate.Value;
                return (StartDiscountFactor == 0 || NbYears == 0) ? 0 : ((1 / EndDiscountFactor / StartDiscountFactor) - 1) / NbYears;
            }
            set
            {
                _userForwardRate = value;
            } 
        }

        public double CashFlow { get; set; }

        public void CalculateDiscountFactor(IList<ZeroCouponDataPoint> discountCurve)
        {
            var cs = mn.CubicSpline.InterpolateNatural(discountCurve.Select(x => x.EndDate.ToOADate()), discountCurve.Select(y => y.DiscountFactor));
            StartDiscountFactor = cs.Interpolate(StartDate.ToOADate());
            EndDiscountFactor = cs.Interpolate(EndDate.ToOADate());
        }

        public void CalculateCashFlow(double notional, double spread = 0.0, bool maturity = false)
        {
            var couponRate = ForwardRate + spread;
            var redemption = maturity ? 1 : 0;
            CashFlow = (redemption + (couponRate * NbYears)) * notional;
        }
    }
}
